src/exe/Main.hs

```diff
@@ -6,147 +6,279 @@ module Main
 
 -- | The main function of the command line interface.
 --
+--   Runs the 'application' and interprets all unhandled effects.
+main :: IO ()
+main =
+  runFinal
+    . embedToFinal
+    . cancelToExit
+    . reportToHandleOrCancel stderr
+    . exceptionToReport exceptionToMessage
+    $ application
+ where
+  exceptionToMessage :: SomeException -> Message
+  exceptionToMessage e = Message Internal (displayException e)
+
+-- | The main computation of the command line interface.
+--
 --   Parses the command line arguments and input file. The transformation is
 --   applied on the parsed input module and a state constructed from the
 --   command line arguments. The output is either printed to the console
 --   or a file.
-main :: IO ()
-main = do
-  args      <- getArgs
-  (opts, _) <- compilerOpts args
-  if not (showHelp opts)
-    then do
-      let state = transformOptions opts
-      input <- readFile $ inputFile opts
-      let x = HSE.fromParseResult (HSE.parseModule input)
-          m = evalPM (processModule (void x)) state
-      case outputDir opts of
-        Just out -> do
-          -- TODO this looks to me as if 'outputDir' is named incorrectly.
-          -- It is not an output directory but the name of the output file.
-          writeFile out (pPrint m)
-          printDebug (enableDebug opts) state
-        Nothing -> do
-          putStr $ pPrint m
-          printDebug (enableDebug opts) state
-    else putStr (usageInfo "" options)
-
--- | Prints the 'debugOutput' from the given 'PMState' to the consolve if
---   the first argument is set to @True@.
-printDebug :: Bool -> PMState -> IO ()
-printDebug b s | b         = print $ "DebugOutput:" ++ debugOutput s
-               | -- TODO pretty debug
-                 otherwise = return ()
+application :: Members '[Report, Embed IO] r => Sem r ()
+application = do
+  args <- embed getArgs
+  opts <- parseArgs args
+
+  -- Filter reported message based on @--debug@ flag.
+  filterReportedMessages
+      (\msg -> optEnableDebug opts || msgSeverity msg /= Debug)
+    $ -- Show usage information when the @--help@ flag is specified or there is no
+      -- input file.
+      if optShowHelp opts || null (optInputFiles opts)
+        then embed putUsageInfo
+        else mapM_ (processInputFile opts) (optInputFiles opts)
+
+-------------------------------------------------------------------------------
+-- Pattern Matching Compilation                                              --
+-------------------------------------------------------------------------------
+
+-- | Applies the transformation to the given file and writes the transformed
+--   module to the console or an output file depending on the command line
+--   options.
+--
+--   The name of the output file is generated by appending the result of
+--   'makeOutputFileName' for the input module to output directory.
+--   If any parent directory of the output file does not exist, it is created.
```

Comment:
Since the first argument of [`createDirectoryIfMissing`](https://hackage.haskell.org/package/directory-1.3.6.1/docs/System-Directory.html#v:createDirectoryIfMissing) is set to `True`, parent directories of the output directory are created as well if they are missing. I think that it is important to document that this is the intended behavior. Maybe the sentence can be improved as follows.

```suggestion
--   'makeOutputFileName' for the input module to the output directory.
--   If the output directory does not exist, the output directory and all of
--   its parent directories are created.
```